### PR TITLE
Update truffle-hdwallet-provider (docs & codebase)

### LINKED ISSUE #1746: truffle-hdwallet-provider error
- [x] I've asked for help in the [Truffle Gitter](http://gitter.im/Consensys/truffle) before filing this issue.
code:
```
const Web3 = require('web3');
const {interface, bytecode} = require('./compile');
const HDWalletProvider = require('truffle-hdwallet-provider');
const mnemonic = "salmon ***"; // 12 word mnemonic
const provider = new HDWalletProvider(mnemonic,
    "https://rinkeby.infura.io/v3/***");
const web3 = new Web3(provider);

deploy = async ()=>{
     const accounts = await web3.eth.getAccounts();
     console.log(accounts[0]);
}
deploy();
```

## Issue

```
node deploy.js
(node:13036) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'map' of undefined
    at GetAccountsMethod.afterExecution (H:\\SmartContracts\Index\node_modules\web3-core-method\dist\web3-core-metho
d.cjs.js:762:23)
    at GetAccountsMethod._callee$ (H:\SmartContracts\Index\node_modules\web3-core-method\dist\web3-core-method.cjs.
js:457:39)
...
...
...
TypeError: e is not a function
    at H:\SmartContracts\Index\node_modules\truffle-hdwallet-provider\dist\index.js:15:544968
    at H:\SmartContracts\Index\node_modules\truffle-hdwallet-provider\dist\index.js:1:150953
    at d (H:\SmartContracts\Index\node_modules\truffle-hdwallet-provider\dist\index.js:1:215490)
    at H:\SmartContracts\Index\node_modules\truffle-hdwallet-provider\dist\index.js:1:215543

```

### PR DESCRIPTION
**Pull request by CruzMolina**
This PR updates `truffle-hdwallet-provider`'s documentation (clarifies current web3.js compatibility version requirement & updates truffle project example to match infura's updated API), upgrades `web3-engine-provider` dep to latest, & adds a shim to support `sendAsync` methods. Also adds `web3@1.0.0-beta.37` as a dependency to temporarily relieve friction between users & current web3 compatibility issues.

Potentially resolves #1746, #1643, #852, & #1874.

### CODE REVIEW

File: packages/truffle-hdwallet-provider/README.md [hunk 1]
```diff
@@ -94,7 +96,7 @@ module.exports = {
     ropsten: {
       // must be a thunk, otherwise truffle commands may hang in CI
       provider: () =>
-        new HDWalletProvider(mnemonic, "https://ropsten.infura.io/"),
+        new HDWalletProvider(mnemonic, "https://ropsten.infura.io/v3/YOUR-PROJECT-ID"),
```
Comment (gnidan) [inline packages/truffle-hdwallet-provider/README.md]: 👌

File: packages/truffle-hdwallet-provider/src/index.js [hunk 2]
```diff
@@ -116,6 +116,8 @@ function HDWalletProvider(
 
   this.engine.addProvider(new FiltersSubprovider());
   if (typeof provider === "string") {
+    Web3.providers.HttpProvider.prototype.sendAsync =
+      Web3.providers.HttpProvider.prototype.send;
```
Comment (gnidan) [inline packages/truffle-hdwallet-provider/src/index.js]: Could we just get rid of `sendAsync` instead of patching it in?

Comment (gnidan) [reply packages/truffle-hdwallet-provider/src/index.js]: If this is necessary for reasons specified in one of the four linked issues, I apologize for being lazy.

Comment (CruzMolina) [reply packages/truffle-hdwallet-provider/src/index.js]: It's necessary for at least one of the linked issues, plus now `web3-provider-engine` throws an err if a provider (or maybe subProvider?) doesn't have that method.

**Comment by CruzMolina** [comment]
Refactored the code to make it easier to reason about in the future. Also documented the Web3 shim.

File: packages/truffle-hdwallet-provider/README.md [hunk 3]
```diff
@@ -10,26 +10,28 @@ $ npm install truffle-hdwallet-provider
 ## Requirements
 ```
 Node >= 7.6
+Web3 <= 1.0.0-beta.37
```
Comment (gnidan) [inline packages/truffle-hdwallet-provider/README.md]: gotta be exactly that!

**Pull request merged**